### PR TITLE
Performance improvement for `cupy.var` complex inputs

### PR DESCRIPTION
**Pull request by asi1024**
Rel #2411.
Speed up `cupy.var` for complex inputs by defining custom kernel for complex value.

### CODE REVIEW

**Comment by niboshi** [comment]
Can you provide benchmark results?

File: cupy/core/_routines_statistics.pyx [hunk 1]
```diff
@@ -234,49 +234,43 @@ cdef _nanargmax = create_reduction_func(
...
 
+    dtype_mean = a.dtype.char
     dtype_out = dtype
-    if dtype in ['D', 'F']:
-        if a.dtype.kind == 'f':
-            dtype = a.dtype.char
+    if dtype is None:
+        if a.dtype.kind in 'biu':
+            dtype_mean = 'd'
+            dtype_out = 'd'
```
Comment (niboshi) [inline cupy/core/_routines_statistics.pyx]: You should not use dtype char code.

https://docs.scipy.org/doc/numpy/reference/arrays.scalars.html#built-in-scalar-types

File: cupy/core/_routines_statistics.pyx [hunk 2]
```diff
@@ -286,15 +280,26 @@ cdef ndarray _std(
...
+__device__ float my_norm(const complex<float>& x) { return norm(x); }
+__device__ double my_norm(const complex<double>& x) { return norm(x); }
+'''
+
+
 cdef _var_core = ReductionKernel(
     'S x, T mean, T alpha', 'T out',
-    '(x - mean) * (x - mean)',
-    'a + b', 'out = alpha * a', '0', '_var_core')
+    'my_norm(static_cast<T>(x) - mean)',
```
Comment (niboshi) [inline cupy/core/_routines_statistics.pyx]: Is this optimization benefitial?
I think this kind of optimization is commonly done by the compiler.

Comment (niboshi) [reply cupy/core/_routines_statistics.pyx]: Sorry, never mind.
I overlooked complex variants use `std::norm`.

**Comment by grlee77** [comment]
> Can you provide benchmark results?

This PR seems to be almost a factor of two improvement over current master.

Using `cupy.std` on a complex64 arrays:

with 100,000 elements:
```
    master: 314 µs ± 12.4 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
    this PR: 168 µs ± 12.4 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

with 100,000,000 elements:
```
    master: 205 ms ± 55.4 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
    this PR: 113 ms ± 8.47 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Similar improvement occurs for `cupy.var`

**Comment by asi1024** [comment]
I removed redundant memory copies of the output array and measured performances with the following 2 types of inputs.
- var_float: `shape=(1000, 1000), dtype='float32'`
- var_complex: `shape=(1000, 1000), dtype='complex64'`

- `master` (`3d5d5d3c4`)
```
var_float      :    49.627 us   +/-53.548 (min:   46.125) us    806.001 us   +/-48.474 (min:  783.776) us
var_complex    :   118.108 us   +/-18.705 (min:  112.390) us   3062.695 us   +/-14.864 (min: 3044.384) us
```

- This PR (`db13a973e`)
```
var_float      :    58.410 us   +/- 8.308 (min:   56.412) us    807.499 us   +/-25.925 (min:  791.104) us
var_complex    :    59.996 us   +/- 9.603 (min:   57.613) us   1677.435 us   +/- 9.098 (min: 1653.408) us
```

**Comment by asi1024** [comment]
Add a commit for more performance improvement.
`HEAD` (`eef646801`)
```
var_float      :    54.315 us   +/- 7.526 (min:   52.076) us    791.380 us   +/- 6.337 (min:  772.160) us
var_complex    :    56.022 us   +/- 8.459 (min:   53.555) us   1671.560 us   +/- 7.467 (min: 1657.728) us
```

Re-measured in `master` because the previous `var_float` result have too large divergence (`+/-53.548`)
```
var_float      :    54.669 us   +/- 8.657 (min:   49.902) us    792.045 us   +/- 6.150 (min:  772.128) us
var_complex    :   130.345 us   +/-15.724 (min:  122.815) us   3064.546 us   +/-10.047 (min: 3040.160) us
```

**Comment by niboshi** [comment]
@asi1024 @grlee77
Thank you for the benchmark results.

@asi1024 Please fix the test failure.

**Comment by asi1024** [comment]
@niboshi PTAL.

**Comment by niboshi** [comment]
Jenkins, test this please

**Pull request merged**